### ISSUE TITLE
Invariants Mining recall and precision

### ISSUE THREAD

**Issue by riyu94** [logpai/loglizer#40]
Hi,

I am using Invariants Mining without Label code to train data and have modified the code to get the anomaly.csv from the output we get from trained data.

Then, I tried using the above generated anomaly.csv instead of the anomaly provided in the project, to test data with train-ratio as 0.5. (using Invariants mining with label)

I get the results (number of anomalies in train and test), but it shows recall and precision as 0.00.
Tried modifying HDFS log file but does not help.

Please help, The code is indeed helpful but documentation needed.
Also, can I deploy this algorithm with Apache Spark? Since it will be too slow otherwise.

**Comment by amineebenamor** [comment]
Hello,
I've already used Invariants Mining and it works for me. Your message isn't enough for me to know where the problem is, as you did some changes to the original code.
If I can have a look at your code, I can check where the error comes from.

**Comment by riyu94** [comment]
Hi

The algorithm works fine but output for precision and recall is not correct since I am generating the anomaly.csv by running the alogrithm.
Below is the code I updated in demo file: training phase

import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_label.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    # Load structured log without label info
    (x_train, _), (x_test, _), (x_data, _) = dataloader.load_HDFS(struct_log,
                                                     window='session', 
                                                     train_ratio=1,
                                                     split_type='sequential')
    print x_data
    # print x_train
    # Feature extraction
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train)
    # print x_train
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)

    # Predict anomalies on the training set offline, and manually check for correctness
    y_train = model.predict(x_train)

    # print y_train

    x_data['Label'] = np.array(y_train)
    # print x_data

    anomaly = pd.DataFrame(columns = ['BlockId','Label'])
    anomaly.BlockId=x_data.BlockId
    anomaly.loc[x_data.Label == 0 , 'Label'] = 'Normal'
    anomaly.loc[x_data.Label == 1 , 'Label'] = 'Anomaly'
    #print anomaly

    anomaly.to_csv('../data/HDFS/anomaly_output.csv', sep=',')

Getting the original dataframe using data_df in dataloader part (passing as an extra argument):


        if label_file is None:
            if split_type == 'uniform':
                split_type = 'sequential'
                print('Warning: Only split_type=sequential is supported \
                if label_file=None.'.format(split_type))
            # Split training and validation set sequentially
            x_data = data_df['EventSequence'].values
            (x_train, _), (x_test, _) = _split_data(x_data, train_ratio=train_ratio, split_type=split_type)
            print('Total: {} instances, train: {} instances, test: {} instances'.format(
                  x_data.shape[0], x_train.shape[0], x_test.shape[0]))
            return (x_train, None), (x_test, None), (data_df, None)

And then, in the testing phase:

import sys
sys.path.append('../')
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_output.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    (x_train, y_train), (x_test, y_test) = dataloader.load_HDFS(struct_log,
                                                                label_file=label_file,
                                                                window='session', 
                                                                train_ratio=0.5,
                                                                split_type='sequential')
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train)
    x_test = feature_extractor.transform(x_test)

    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)
    
    print('Train validation:')
    precision, recall, f1 = model.evaluate(x_train, y_train)
    
    print('Test validation:')
    precision, recall, f1 = model.evaluate(x_test, y_test)


Attaching the generated anomaly_output.csv copy as excel file, for reference.
[Copy of anomaly_output.xlsx](https://github.com/logpai/loglizer/files/3128069/Copy.of.anomaly_output.xlsx)

**Comment by riyu94** [comment]
below if the output i get for testing phase:


====== Input data summary ======
Total: 7940 instances, 150 anomaly, 7790 normal
Train: 3970 instances, 68 anomaly, 3902 normal
Test: 3970 instances, 82 anomaly, 3888 normal

====== Transformed train data summary ======
Train data shape: 3970-by-14

====== Transformed test data summary ======
Test data shape: 3970-by-14

====== Model summary ======
Invariant space dimension: 11
Mined 10 invariants: {(1, 13): [1.0, -1.0], (3, 11): [-2.0, 1.0], (0, 12): [-10.0, 1.0], (1, 8): [-1.0, 1.0], (3, 9): [-101.0, 1.0], (1, 6): [1.0, -3.0], (1, 10): [-1.0, 1.0], (3, 7): [1.0, -1.0], (3, 4): [1.0, -1.0], (0, 2): [1.0, -1.0]}

Train validation:
====== Evaluation summary ======
Precision: 0.000, recall: 0.000, F1-measure: 0.000

Test validation:
====== Evaluation summary ======
Precision: 0.000, recall: 0.000, F1-measure: 0.000

**Comment by riyu94** [comment]
#!/usr/bin/env python
# -*- coding: utf-8 -*-
''' This is a demo file for the Invariants Mining model.
    API usage:
        dataloader.load_HDFS(): load HDFS dataset
        feature_extractor.fit_transform(): fit and transform features
        feature_extractor.transform(): feature transform after fitting
        model.fit(): fit the model
        model.predict(): predict anomalies on given data
        model.evaluate(): evaluate model accuracy with labeled data
'''

import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_output.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    # Load structured log without label info
    (x_train, _), (x_test, _), (x_data, _) = dataloader.load_HDFS(struct_log,
                                                     window='session', 
                                                     train_ratio=1,
                                                     split_type='sequential')
    #print x_data
    # print x_train
    # Feature extraction
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train)
    # print x_train
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)

    # Predict anomalies on the training set offline, and manually check for correctness
    y_train = model.predict(x_train)

    # print y_train

    x_data['Label'] = np.array(y_train)
    # print x_data

    anomaly = pd.DataFrame(columns = ['BlockId','Label'])
    anomaly.BlockId=x_data.BlockId
    anomaly.loc[x_data.Label == 0 , 'Label'] = 'Normal'
    anomaly.loc[x_data.Label == 1 , 'Label'] = 'Anomaly'
    #print anomaly

    anomaly.to_csv('../data/HDFS/anomaly_output.csv', sep=',')
    # Predict anomalies on the test set to simulate the online mode
    # x_test may be loaded from another log file
    x_test = feature_extractor.transform(x_test)
    y_test = model.predict(x_test)

    # print x_test
    # print y_test
    # # If you have labeled data, you can evaluate the accuracy of the model as well.
    # # Load structured log with label info
    (x_train, y_train), (x_test, y_test) = dataloader.load_HDFS(struct_log,
                                                                label_file=label_file,
                                                                window='session', 
                                                                train_ratio=0.5,
                                                                split_type='sequential')   
    x_test = feature_extractor.transform(x_test)
    precision, recall, f1 = model.evaluate(x_test, y_test)


Output:

====== Input data summary ======
~Total: 7940 instances, train: 7940 instances, test: 0 instances
====== Transformed train data summary ======
Train data shape: 7940-by-19

====== Model summary ======
Invariant space dimension: 16
Mined 16 invariants: {(1, 18): [1.0, -1.0], (6, 8): [1.0, -1.0], (0, 15): [-35.0, 1.0], (1, 14): [-1.0, 1.0], (5, 13): [-101.0, 1.0], (0, 17): [1.0, -2.0], (0, 6): [-17.0, 1.0], (0, 16): [-9.0, 1.0], (0, 5): [-17.0, 1.0], (0, 10): [-17.0, 1.0], (6, 13): [-101.0, 1.0], (0, 4): [1.0, -1.0], (0, 3): [-1.0, 1.0], (1, 11): [-1.0, 1.0], (5, 8): [1.0, -1.0], (1, 9): [1.0, -3.0]}

====== Transformed test data summary ======
Test data shape: 0-by-19

====== Input data summary ======
Total: 7940 instances, 150 anomaly, 7790 normal
Train: 3970 instances, 71 anomaly, 3899 normal
Test: 3970 instances, 79 anomaly, 3891 normal

====== Transformed test data summary ======
Test data shape: 3970-by-19

====== Evaluation summary ======
Precision: 0.018, recall: 0.025, F1-measure: 0.021


When I run the test data after generating the anomaly_output.csv in the same script,
I get the above output.

Basically using the anomaly.csv generated from data, doesn't give good results.
How can I not use the provided anomaly file and get good results generating a new anomaly.csv with the trained data?

**Comment by riyu94** [comment]
Works 💃 
sorry for the confusion :)

**Comment by riyu94** [comment]
Made some changes and it gave the precision and recall as 100% as I am using same data for training and testing, that was expected.
The below code works well:
#!/usr/bin/env python
# -*- coding: utf-8 -*-
''' This is a demo file for the Invariants Mining model.
    API usage:
        dataloader.load_HDFS(): load HDFS dataset
        feature_extractor.fit_transform(): fit and transform features
        feature_extractor.transform(): feature transform after fitting
        model.fit(): fit the model
        model.predict(): predict anomalies on given data
        model.evaluate(): evaluate model accuracy with labeled data
'''

import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_output.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    # Load structured log without label info
    (x_train, _), (x_test, _), (x_data, _) = dataloader.load_HDFS(struct_log,
                                                     window='session', 
                                                     train_ratio=1,
                                                     split_type='sequential')
    # Feature extraction
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train) #transform train data
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)  #invariant space and search

    # Predict anomalies on the training set offline, and manually check for correctness
    y_train = model.predict(x_train)

    x_data['Label'] = np.array(y_train)

    # get the anomaly label file
    anomaly = pd.DataFrame(columns = ['BlockId','Label'])
    anomaly.BlockId=x_data.BlockId
    anomaly.loc[x_data.Label == 0 , 'Label'] = 'Normal'
    anomaly.loc[x_data.Label == 1 , 'Label'] = 'Anomaly'
    #print anomaly

    anomaly.to_csv('../data/HDFS/anomaly_output.csv', sep=',')

    # Predict anomalies on the test set to simulate the online mode

    # # If you have labeled data, you can evaluate the accuracy of the model as well.
    # # Load structured log with label info
    (x_train, y_train), (x_test, y_test) = dataloader.load_HDFS(struct_log,
                                                                label_file=label_file,
                                                                window='session', 
                                                                train_ratio=0.5,
                                                                split_type='sequential')   
    x_test = feature_extractor.transform(x_test)
    y_test = model.predict(x_test)
    precision, recall, f1 = model.evaluate(x_test, y_test)


Output:

====== Input data summary ======
Total: 7940 instances, train: 7940 instances, test: 0 instances
====== Transformed train data summary ======
Train data shape: 7940-by-19

====== Model summary ======
Invariant space dimension: 16
Mined 16 invariants: {(1, 18): [1.0, -1.0], (6, 8): [1.0, -1.0], (0, 15): [-35.0, 1.0], (1, 14): [-1.0, 1.0], (5, 13): [-101.0, 1.0], (0, 17): [1.0, -2.0], (0, 6): [-17.0, 1.0], (0, 16): [-9.0, 1.0], (0, 5): [-17.0, 1.0], (0, 10): [-17.0, 1.0], (6, 13): [-101.0, 1.0], (0, 4): [1.0, -1.0], (0, 3): [-1.0, 1.0], (1, 11): [-1.0, 1.0], (5, 8): [1.0, -1.0], (1, 9): [1.0, -3.0]}

====== Input data summary ======
Total: 7940 instances, 150 anomaly, 7790 normal
Train: 3970 instances, 81 anomaly, 3889 normal
Test: 3970 instances, 69 anomaly, 3901 normal

====== Transformed test data summary ======
Test data shape: 3970-by-19

====== Evaluation summary ======
Precision: 1.000, recall: 1.000, F1-measure: 1.000

Here, Now I need help with some testing data, what changes I can make in the log to use the generated anomaly.csv as input and get different results for different testing data.
Please help with this.
Thanks

**Comment by riyu94** [comment]
This is my code, where I want to generate the anomaly.csv using train data as train_ratio:1 and provide it as an input when testing data as train_ratio: 0.5
`#!/usr/bin/env python
# -*- coding: utf-8 -*-
''' This is a demo file for the Invariants Mining model.
    API usage:
        dataloader.load_HDFS(): load HDFS dataset
        feature_extractor.fit_transform(): fit and transform features
        feature_extractor.transform(): feature transform after fitting
        model.fit(): fit the model
        model.predict(): predict anomalies on given data
        model.evaluate(): evaluate model accuracy with labeled data
'''

import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_output.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    # Load structured log without label info
    (x_train, _), (x_test, _), (x_data, _) = dataloader.load_HDFS(struct_log,
                                                     window='session', 
                                                     train_ratio=1,
                                                     split_type='sequential')
    # Feature extraction
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train) #transform train data
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)  #invariant space and search

    # Predict anomalies on the training set offline, and manually check for correctness
    y_train = model.predict(x_train)

    x_test = feature_extractor.transform(x_test)
    y_test = model.predict(x_test)

    x_data['Label'] = np.array(y_train)

    # get the anomaly label file
    anomaly = pd.DataFrame(columns = ['BlockId','Label'])
    anomaly.BlockId=x_data.BlockId
    anomaly.loc[x_data.Label == 0 , 'Label'] = 'Normal'
    anomaly.loc[x_data.Label == 1 , 'Label'] = 'Anomaly'
    # print anomaly

    anomaly.to_csv('../data/HDFS/anomaly_output.csv', sep=',')

    # Predict anomalies on the test set to simulate the online mode

    # If you have labeled data, you can evaluate the accuracy of the model as well.
    # Load structured log with label info
    (x_train, y_train), (x_test, y_test) = dataloader.load_HDFS(struct_log,
                                                                label_file=label_file,
                                                                window='session', 
                                                                train_ratio=0.5,
                                                                split_type='sequential')   
    x_test = feature_extractor.transform(x_test)
    precision, recall, f1 = model.evaluate(x_test, y_test)

`

Output:

<img width="824" alt="Capture" src="https://user-images.githubusercontent.com/21206708/56920400-f5dc0380-6a77-11e9-80b7-fac92f93ad76.PNG">


Very low precision and recall though i am using same log file for training and testing.
If I am doing something worng with using the same log file, what changes should i make to the input?
Please help

**Comment by riyu94** [comment]
Attach

**Comment by riyu94** [comment]
Code:

#!/usr/bin/env python
# -*- coding: utf-8 -*-
''' This is a demo file for the Invariants Mining model.
    API usage:
        dataloader.load_HDFS(): load HDFS dataset
        feature_extractor.fit_transform(): fit and transform features
        feature_extractor.transform(): feature transform after fitting
        model.fit(): fit the model
        model.predict(): predict anomalies on given data
        model.evaluate(): evaluate model accuracy with labeled data
'''

import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
label_file = '../data/HDFS/anomaly_output.csv' # The anomaly label file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    #Load structured log without label info
    (x_train, _), (x_test, _), (x_data, _) = dataloader.load_HDFS(struct_log,
                                                     window='session', 
                                                     train_ratio=1,
                                                     split_type='sequential')
    # Feature extraction
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train) #transform train data
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)  #invariant space and search

    # Predict anomalies on the training set offline, and manually check for correctness
    y_train = model.predict(x_train)

    x_data['Label'] = np.array(y_train)

    # get the anomaly label file
    anomaly = pd.DataFrame(columns = ['BlockId','Label'])
    anomaly.BlockId=x_data.BlockId
    anomaly.loc[x_data.Label == 0 , 'Label'] = 'Normal'
    anomaly.loc[x_data.Label == 1 , 'Label'] = 'Anomaly'
    # print anomaly

    anomaly.to_csv('../data/HDFS/anomaly_output.csv', sep=',')

    # Predict anomalies on the test set to simulate the online mode

    # If you have labeled data, you can evaluate the accuracy of the model as well.
    # Load structured log with label info
    (x_train, y_train), (x_test, y_test) = dataloader.load_HDFS(struct_log,
                                                                label_file=label_file,
                                                                window='session', 
                                                                train_ratio=0.5,
                                                                split_type='sequential')   

    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train) #transform train data
    # Model initialization and training
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)  #invariant space and search
    x_test = feature_extractor.transform(x_test)
    print('Train validation:')
    precision, recall, f1 = model.evaluate(x_train, y_train)
    
    print('Test validation:')
    precision, recall, f1 = model.evaluate(x_test, y_test)

Output:

====== Input data summary ======
Total: 7940 instances, train: 7940 instances, test: 0 instances
====== Transformed train data summary ======
Train data shape: 7940-by-19

====== Model summary ======
Invariant space dimension: 16
Mined 16 invariants: {(1, 18): [1.0, -1.0], (6, 8): [1.0, -1.0], (0, 15): [-35.0, 1.0], (1, 14): [-1.0, 1.0], (5, 13): [-101.0, 1.0], (0, 17): [1.0, -2.0], (0, 6): [-17.0, 1.0], (0, 16): [-9.0, 1.0], (0, 5): [-17.0, 1.0], (0, 10): [-17.0, 1.0], (6, 13): [-101.0, 1.0], (0, 4): [1.0, -1.0], (0, 3): [-1.0, 1.0], (1, 11): [-1.0, 1.0], (5, 8): [1.0, -1.0], (1, 9): [1.0, -3.0]}

====== Input data summary ======
Total: 7940 instances, 150 anomaly, 7790 normal
Train: 3970 instances, 84 anomaly, 3886 normal
Test: 3970 instances, 66 anomaly, 3904 normal

====== Transformed train data summary ======
Train data shape: 3970-by-14

====== Model summary ======
Invariant space dimension: 11
Mined 10 invariants: {(1, 13): [1.0, -1.0], (3, 11): [-2.0, 1.0], (0, 12): [-10.0, 1.0], (1, 8): [-1.0, 1.0], (3, 9): [-101.0, 1.0], (1, 6): [1.0, -3.0], (1, 10): [-1.0, 1.0], (3, 7): [1.0, -1.0], (3, 4): [1.0, -1.0], (0, 2): [1.0, -1.0]}

====== Transformed test data summary ======
Test data shape: 3970-by-14

Train validation:
====== Evaluation summary ======
Precision: 0.000, recall: 0.000, F1-measure: 0.000

Test validation:
====== Evaluation summary ======
Precision: 0.009, recall: 0.015, F1-measure: 0.011

In the first part, I train the data by using ratio 1 and generate the anomal.csv file
After that, I use ratio 0.5 to train and test and evaluate the results using the generated anomaly.csv
And the results come out to be pretty low.

Shouldn't the output evaluation been much more, since I am using the output  from trained file to compare the results.
Please guide me with the same.
Thank you

**Comment by amineebenamor** [comment]
Did you solve your problem? What did you do then?
I didn't have time to look for it today I was going to see it tomorrow.

**Comment by riyu94** [comment]
Yes, I was assuming that the anomaly.csv generated by keeping the train ratio as 1, can be used to evaluate results when we test the same input log using train ratio 0.5. But in both cases, mining invariants list was different so It will not be the same.

Anyway, no problem. I had another question, Invariants Mining, what kind of anomalies it is good for? 1. Like if we have new type of (never seen) log messages, will it be able to detect?
2. Or if we have a sudden spike in a particular type of message, will it be able to mark it?

**Comment by amineebenamor** [comment]
The Invariants Mining algorithm extracts invariants from the training set: for example, in the normal executions of a system, the number of log messages indicating "Open file" is usually equal to the number of log messages corresponding to "Close file", because each opened file will be closed at some stage eventually. These invariants represents the normal execution flows of the system. It also gives an easy interpretation for each anomaly as we can check which invariant is broken.
Therefore, it is good for any type of anomaly that would break the normal execution of the system (by breaking the invariants).
To answer your question:
1. If a new log breaks certain invariants, we detect an anomaly that occured during the system execution. Otherwise, it won't be specified as an anomaly.
2. It depends on the type of sudden spike in a particular type of message, but if it breaks an invariant, then it means that it has broken a normal execution of the system and will be considered as an anomaly.
Let me know if you have more questions, I'm working on Invariants Mining also :)

**Comment by riyu94** [comment]
@amineebenamor  Helps :)
Thank you.

**Comment by jimzhu** [comment]
@amineebenamor @riyu94   Many thanks for amineebenamor‘ support! I want to make some additional comments to amineebenamor‘ answer. 
> 1. Like if we have new type of (never seen) log messages, will it be able to detect?
Yes, our current implementation supports the detection of this type of anomaly. In the [fit_transform method](https://github.com/logpai/loglizer/blob/master/loglizer/preprocessing.py#L28), we add two arguments `oov=False` and `min_count=1`. When setting oov=True during preprocessing, we will mark all events < min_count as 'oov' events (out-of-vocabulary). When setting min_count=1, we get normal instances with zero oov events, but new events will make oov events>1. This matches your problem. Meanwhile, we can also drop some rarely occur events during training by setting, for example, min_count=5.
> 2. Or if we have a sudden spike in a particular type of message, will it be able to mark it?
If one event type gets a sudden spike, it will lead to an invariant break. But a sudden spike of an event pair will not.
